src/NuGet.Core/NuGet.Common/PathUtil/DirectoryUtility.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Directory operation helpers.
+    /// </summary>
+    public static class DirectoryUtility
+    {
+        private static object s_sharedFolderCreationLock = new object();
```

Comment:
This is only process-wide. Given that a machine-wide lock is sorta the higher level goal here, can we make this code resilient to failures? That is, if two processes start this at the same time, one will have its `Directory.CreateDirectory` fail with a specific exception, right?